nixops/resources/vault_approle.py

```diff
@@ -0,0 +1,218 @@
+import nixops.util
+import nixops.resources
+import nixops.vault_common
+import json
+import requests
+
+
+class VaultApproleDefinition(nixops.resources.ResourceDefinition):
+    """Definition of a Vault Approle"""
+
+    @classmethod
+    def get_type(cls):
+        return "vault-approle"
+
+    @classmethod
+    def get_resource_type(cls):
+        return "vaultApprole"
+
+    def show_type(self):
+        return "{0}".format(self.get_type())
+
+
+class VaultApproleState(nixops.resources.DiffEngineResourceState):
+    """State of a Vault Approle"""
+
+    state = nixops.util.attr_property("state", nixops.resources.ResourceState.MISSING, int)
+    vault_token = nixops.util.attr_property("vaultToken", None)
+    _reserved_keys = [ "vaultAddress", "roleName", "roleId", "bindSecretId"
```

Comment:
are these really `_reserved_keys` ? it seems like these are actually the options from nix expression that defines the vault app role so you actually need them to fire up the diff engine.

`_reserved_keys` is meant to be used for ignoring attributes that are stored or internal which we don't want to handle e.g in EC2 the instance id would be reserved as otherwise the diff engine will always notice that there is a diff between the config and the state.

Reply:
@AmineChikhaoui  ok so `_reserved` is used for attributes set by nixops while the options are set using `nixops.util.attr_property` ?